### ISSUE TITLE
File dialogue not opening programmatically on older iOS

### ISSUE THREAD

**Issue by einsiol** [react-dropzone/react-dropzone#1009]
**Do you want to request a *feature* or report a *bug*?**

- [x] I found a bug
- [ ] I want to propose a feature

**What is the current behavior?**
Using useDropzone hooks

File dialogue cannot be programmatically opened by using the click method on the file input.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Create a function (in this case onTap handler for a parent [Hammer component](https://github.com/JedWatson/react-hammerjs)) and use the inputRef to have a reference to the file input. The function should be a call to inptuRef.current.click() to open the file dialog
1. Overwrite default styles of the input by making the input a `display:block`, then also adding `visibility: hidden` and _width_ and _height_ to _0px_.
1. Run the function to open the file dialogue.

**What is the expected behavior?**
When calling the click method programmatically for the File Input, it should open the file dialogue on Safari on iOS < 12.4

It works without an issue on all other browsers and on android.

**Please mention other relevant information such as the browser version, Operating System and react-dropzone version.**
 iOS < 12.4 (tested on an iPhone 6 and iPad with build number a1474)
react-dropzone v11.0.2

I don't have access to devices to test this directly myself, seems that xCode Simulators do not have the same issue. This issue was reported by the QA testers on my project.

**Comment by rolandjitsu** [comment]
@einsiol you can imagine that it would be difficult for us to test this on actual hardware with that version of Safari.

But something else could be going on here. Since you use the hammer lib, could that be interfering with the click event?

Also, we don't handle [touch events](https://reactjs.org/docs/events.html#touch-events), so it's possible that the click event doesn't get triggered.

Would be great if you could look into this using the hardware where this isn't working since we cannot reproduce in simulators.

**Comment by einsiol** [comment]
Hey, @rolandjitsu So I have had the opportunity to confirm a few things. The touch event is indeed handled by Hammer.js, I can confirm that the tab event is triggered correctly on the iOS 12 devices. So the handler for the touch event is triggered correctly. I had a user of an iOS 12 device test this by having an alert firing when the event was triggered.

 I've used both a ref to reference directly to the input and called inputRef.current.click() to open the input dialog, this failed, I've also tried to use the open() function that comes from the hook. Both don't look like they are not opening the file dialog when called.

On all other devices and main browsers, there are no issues.

This might not, of course, strictly speaking, be a bug in React-Dropzone. But I'm at a loss and can't find a solution for this in wondered if someone had find a solution, it could be integrated into React-Dropzone :)

**Comment by rolandjitsu** [comment]
> Hey, @rolandjitsu So I have had the opportunity to confirm a few things. The touch event is indeed handled by Hammer.js, I can confirm that the tab event is triggered correctly on the iOS 12 devices. So the handler for the touch event is triggered correctly. I had a user of an iOS 12 device test this by having an alert firing when the event was triggered.
> 
> I've used both a ref to reference directly to the input and called inputRef.current.click() to open the input dialog, this failed, I've also tried to use the open() function that comes from the hook. Both don't look like they are not opening the file dialog when called.
> 
> On all other devices and main browsers, there are no issues.
> 
> This might not, of course, strictly speaking, be a bug in React-Dropzone. But I'm at a loss and can't find a solution for this in wondered if someone had find a solution, it could be integrated into React-Dropzone :)

Thanks for the info. I just did a quick search about this and found these threads:
* [manual-trigger-on-inputtype-file-doesnt-work-on-ios-12-2](https://stackoverflow.com/questions/55569608/manual-trigger-on-inputtype-file-doesnt-work-on-ios-12-2)
* https://github.com/danialfarid/ng-file-upload/issues/1766
* https://github.com/flarum/core/issues/1128

There's a mention of using [fastclick](https://github.com/ftlabs/fastclick) which may cause these issues, is this your case?

There's also a mention of needing to use the touch events. Could you test if this is the case by adding a simple [component](https://codesandbox.io/s/trusting-chandrasekhar-joxbx) somewhere and checking if it works:
```jsx
function TestOpenInput() {
  const inputRef = React.createRef();
  const openWithClick = () => {
    inputRef.current.click();
  };

  const openWithTouch = () => {
    inputRef.current.click();
  };

  return (
    <div>
      <button onClick={openWithClick}>Click me</button>
      <button onTouchEnd={openWithTouch}>Touch me</button>
      <input
        type="file"
        ref={inputRef}
        style={{
          position: "absolute",
          width: 0,
          height: 0,
          top: -9999,
          left: -9999
        }}
      />
    </div>
  );
}
```

Let me know if both of the buttons trigger the file select dialog to open.

**Comment by einsiol** [comment]
@rolandjitsu So what I know for sure is that when the user taps/click on the custom button the event is triggered, it's just that the `open()` function from the `useDropzone` hook does not trigger the file dialogue on iOS 12, neither does calling `inputRef.current.click()`

But for some reason, this issue does not happen when using the `ReactDropzoneComponent`, and I finally figured out why. I read somewhere that Safari has some extra security when calling inputs programmatically and if there is any delay in the response it will not trigger the event. (I'm trying to find this information again, will post it here when I do).

In the case of my issue, the input is deeply nested into the DOM. Since react registers events at the very top of the DOM (if I understand correctly) then the bubbling of the event takes to long and the input click will not be triggered.

To solve the problem, I had to use React.CreatePortal to move the input to the very bottom of the `<body>` and make sure it is an as close relative of the body as possible. This fixed the problem.

So basically, appending the input to the `<body>` solved the problem for iOS 12.

**Comment by einsiol** [comment]
Can see some implementation details on a StackOverflow topic I created and answered.

https://stackoverflow.com/questions/63898487/input-click-cant-be-triggered-programanticly-on-ios-12/64099306#64099306

I'll close this issue since it not really an issue with React-Dropzone but a bug in WebKit on iOS 12

**Comment by rolandjitsu** [comment]
@einsiol glad you've figured it out.
